appservice/src/utils/workspace.ts

```diff
@@ -35,18 +35,20 @@ export async function selectWorkspaceFile(context: IActionContext, placeHolder:
             canSelectMany: false,
             openLabel: localize('select', 'Select'),
             filters: filters
-        });
+        },
+        recommendedFiles);
 }
 
-export async function selectWorkspaceItem(context: IActionContext, placeHolder: string, options: vscode.OpenDialogOptions): Promise<string> {
+export async function selectWorkspaceItem(context: IActionContext, placeHolder: string, options: vscode.OpenDialogOptions,
+                                          recommendations?: IAzureQuickPickItem<string>[]): Promise<string> {
     let folder: IAzureQuickPickItem<string | undefined> | undefined;
     if (vscode.workspace.workspaceFolders) {
-        const folderPicks: IAzureQuickPickItem<string | undefined>[] = await Promise.all(vscode.workspace.workspaceFolders.map((f: vscode.WorkspaceFolder) => {
-            return { label: path.basename(f.uri.fsPath), description: f.uri.fsPath, data: f.uri.fsPath };
+        const defaultRecommendations: IAzureQuickPickItem<string | undefined>[] = await Promise.all(vscode.workspace.workspaceFolders.map((f: vscode.WorkspaceFolder) => {
+            return {label: path.basename(f.uri.fsPath), description: f.uri.fsPath, data: f.uri.fsPath};
         }));
-
-        folderPicks.push({ label: localize('azFunc.browse', '$(file-directory) Browse...'), description: '', data: undefined });
-        folder = await context.ui.showQuickPick(folderPicks, { placeHolder });
+        const providedItems: IAzureQuickPickItem<string | undefined>[] = recommendations?.length ? recommendations : defaultRecommendations;
```

Comment:
This might be up for debate, but I don't like the idea of totally replacing the folderPicks with the recommendations.  They feel less like recommendations if I have to use browse to deploy my current workspace.  I would rather that they get unshifted into the current folderPicks array so that they appear at the top, but the other quickPicks are still available.

Also, since you are using this to recommend maven artifacts, have you considered using the `deploySubpath` setting available in App Service to deploy the `.jar` file rather than implementing this recommendation feature?  Would that accomplish what you are trying to do?

Reply:
**I updated to use a "better" word suggestion instead of recommendation.**

I updated part of the code in the latest commit.
```js
const defaultSuggestions: IAzureQuickPickItem<string | undefined>[] = await Promise.all(vscode.workspace.workspaceFolders.map((f: vscode.WorkspaceFolder) => {
    return { label: path.basename(f.uri.fsPath), description: f.uri.fsPath, data: f.uri.fsPath };
}));
suggestions = suggestions?.length ? suggestions : defaultSuggestions;
const fileChooser = { label: localize('azFunc.browse', '$(file-directory) Browse...'), description: '', data: undefined };
folder = await context.ui.showQuickPick([...suggestions, fileChooser], { placeHolder });
```

let's call the action item with label `Browse...`  -> **fileChooser**.
and I would like to call the other value items(except *fileChooser*) shown directly in dropdown box `suggested items`/`suggestions`. Because **they are there only because we think user is likely to select them when deploying**. Which is quite similar to the suggestions when you google.

**I think it's a bug to always show the path of workspace on suggestion box** even if we have implicitly indicated only file can be selected(by specifying the `fileExtensions` of `getDeployFsPath`).
as you know, both `selectWorkspaceFile` and `selectWorkspaceFolder` finally delegates to `selectWorkspaceItem`. but `selectWorkspaceItem` always add the *path of current workspace* into suggestion list.

**`deploySubpath` is used, but not as you suggested.** A maven project may have multiple modules. user can choose any module to deploy. when user deploy from a maven project, they are likely to deploy any maven module in the project, so the suggestion feature will detect all the deployable maven modules and suggest them to user. 
`deploySubpath` is used to remember the default module to deployed. and a `preDeployTask: package` is added to build artifact(realy jar/war file) from the module.